libp2p/autotls/acme.nim

```diff
@@ -0,0 +1,317 @@
+import options, json, base64
+from times import DateTime, parse
+import chronos/apps/http/httpclient, jwt, results, bearssl/pem
+
+import ./utils
+import ../crypto/crypto
+import ../crypto/rsa
+import ../transports/tls/certificate_ffi
+import ../transports/tls/certificate
+
+const
+  LetsEncryptURL* = "https://acme-v02.api.letsencrypt.org"
+  LetsEncryptURLStaging* = "https://acme-staging-v02.api.letsencrypt.org"
+  Alg = "RS256"
+  DefaultChalCompletedRetries = 10
+  DefaultChalCompletedRetryTime = 1.seconds
+  DefaultFinalizeRetries = 10
+  DefaultFinalizeRetryTime = 1.seconds
+  DefaultRandStringSize = 256
+
+type ACMEAccount* = object
+  status*: Opt[string]
+  contact*: Opt[seq[string]]
+  key*: KeyPair
+  session*: HttpSessionRef
+  kid*: Opt[string]
+  directory: JsonNode
+  acmeServerURL: string
+
+proc new*(
+    T: typedesc[ACMEAccount],
+    key: KeyPair,
+    status: Opt[string] = Opt.none(string),
+    contact: Opt[seq[string]] = Opt.none(seq[string]),
+    kid: Opt[string] = Opt.none(string),
+    acmeServerURL: string = LetsEncryptURL,
+): Future[ref ACMEAccount] {.async: (raises: [ACMEError, CancelledError]).} =
+  try:
+    let session = HttpSessionRef.new()
+    let directoryResponse =
+      await HttpClientRequestRef.get(session, acmeServerURL & "/directory").get().send()
+    let directory = bytesToString(await directoryResponse.getBodyBytes()).parseJson()
+
+    let acc = new(ACMEAccount)
+    acc.status = status
+    acc.contact = contact
+    acc.kid = kid
+    acc.key = key
+    acc.session = session
+    acc.directory = directory
+    acc.acmeServerURL = acmeServerURL
+    return acc
```

Comment:
nitpick: consider not having practice of having big `try-except` statements. 
when developers add code, the just add code in place, extending `try-except` even more, then they have `try-except` in `try-except`... 

in this case `ACMEAccount` can be initialized after  `try-except